### ISSUE TITLE
Program takes about 9 minutes to (fail to) parse

### ISSUE THREAD

**Issue by yuriy0** [hakaru-dev/hakaru#78]
`RoundTrip.hs/rmProg4` when converted to concrete syntax takes about 9 minutes to fail to parse (on the second line of a 300 line program):

```
$ time pretty examples/rmProg4.hk
"<input>" (line 2, column 5):
unexpected '='
expecting operator or end of input

real    9m8.713s
user    0m0.000s
sys     0m0.000s
```

the parse error is caused by a misplaced keyword (and this program is generated by Hakaru), but the fixed program doesn't parse in less time.

It seems this is related to a large number of nested `case` and/or `let` statements in the expression.

**Comment by JacquesCarette** [comment]
@zaxtax could you look into this?  9 minutes for a 300 line program seems a tad excessive...

**Comment by zaxtax** [comment]
There is probably some accidentally quadratic bits in the pretty printer.

**Comment by JacquesCarette** [comment]
That would be the most obvious guess -- any suspicions as to where?

**Comment by yuriy0** [comment]
> pretty printer

That should be 'parser' - the program doesn't even reach the pretty printer!

**Comment by zaxtax** [comment]
That's bad since I have even less of a clue what to do there.

**Comment by zachsully** [comment]
I have another program [https://github.com/zachsully/hkc-test/blob/master/tests/hakaru/naryProbSum100.fails](prog) that gets parsed into a large _nested_ nary op. It eats through my 16gb of RAM before failing.

rmProg4.hk doesn't have a memory problem, but parsing problems may relate to deep nestings in general.

**Comment by JacquesCarette** [comment]
(Same guess as Rob, but more specific).  Feels like something is recursing 'on the wrong side'.  Maybe the grammar combinators are left-recursive somewhere, which is the worst case for parsec.

**Comment by yuriy0** [comment]
This was fixed by f7ceefa9a86c74fee03a413b8d803cc50f7aedf5.
